lib/reqd.ml

```diff
@@ -138,29 +136,32 @@ let respond_with_bigstring t response (bstr:Bigstringaf.t) =
   if t.error_code <> `Ok then
     failwith "httpaf.Reqd.respond_with_bigstring: invalid state, currently handling error";
   match t.response_state with
-  | Waiting when_done_waiting ->
+  | Waiting ->
     (* XXX(seliopou): check response body length *)
     Writer.write_response     t.writer response;
     Writer.schedule_bigstring t.writer bstr;
     if t.persistent then
       t.persistent <- Response.persistent_connection response;
     t.response_state <- Complete response;
-    done_waiting when_done_waiting
+    Writer.wakeup t.writer;
   | Streaming _ ->
     failwith "httpaf.Reqd.respond_with_bigstring: response already started"
   | Complete _ ->
     failwith "httpaf.Reqd.respond_with_bigstring: response already complete"
 
 let unsafe_respond_with_streaming ~flush_headers_immediately t response =
   match t.response_state with
-  | Waiting when_done_waiting ->
+  | Waiting ->
     let response_body = Body.create t.response_body_buffer in
     Writer.write_response t.writer response;
-    if not flush_headers_immediately then Writer.yield t.writer;
     if t.persistent then
       t.persistent <- Response.persistent_connection response;
     t.response_state <- Streaming(response, response_body);
-    done_waiting when_done_waiting;
+    Body.when_ready_to_write response_body (fun () ->
```

Comment:
How about just making the callback an argument to `Body.create`?

Reply:
Did this in the last commit I pushed, which exposed the fact that I could have done the same pattern in the client connection side. But also it meant making the interface of `Body` less unified, which I think is maybe more natural anyway.